src/layers/SoftMax.jl

```diff
@@ -1,41 +1,56 @@
 type SoftMax <: Nonlinearity
-    last_input  :: Array{Float64}
-    last_output :: Array{Float64}
-    last_loss   :: Array{Float64}
+    x  :: Array{Float64}        # input m by n matrix which represents m instances with scores for n classes
+    y :: Array{Float64}         # output m by n matrix which uses softmax to normalize the input matrix
+    has_init :: Bool            # true if the layer has been initialized
+    jacobian :: Array{Float64}  # cache for the jacobain matrices used in backward
+    dldx :: Array{Float64}      # cahce for the backward result
 
     function SoftMax()
-        return new(Float64[], Float64[], Float64[])
+        return new(Float64[], Float64[], false, Float64[], Float64[])
     end
 end
 
-function forward(l::SoftMax, X::Array{Float64}; kwargs...)
-	X = broadcast(+, X, -maximum(X))
-    l.last_input = X
-    sumX = sum(exp(X))
-    Y = (x ./ sumX)
-    l.last_output = Y
-    return Y
+function init(l::SoftMax, p::Union{Layer,Void}, config::Dict{String,Any}; kwargs...)
+  if p == nothing
+      @assert ndims(config["input_size"]) == 1 # TODO: maybe an error message?
+      out_size = (config["batch_size"], config["input_size"][1])
+  else
+      out_size = getOutputSize(p)
+  end
+  l.x = Array{Float64}(out_size)
+  l.y = Array{Float64}(out_size)
+  l.has_init = true;
+  m,n = out_size
+  l.jacobian = Array{Float64}(n,n)
+  l.dldx = Array{Float64}(out_size)
+end
+
+function forward(l::SoftMax, X::Array{Float64,2}; kwargs...)
+
+    l.x = X
+    # iterating each row/picture
+    lxexp = exp(l.x)
+    lxsum = sum(lxexp, 2)
+    l.y = lxexp./lxsum
+    return l.y
 
 end
 
 function backward(l::SoftMax, DLDY::Array{Float64}; kwargs...)
-    @assert size(l.last_input) == size(DLDY)
-    X = l.last_input
-    sumX = sum(exp(X))
-    u = zeros(ndims(X), ndims(X))
-    z = zeros(ndims(X))
-    for i = 1: ndims(X)
-        z[i] = (X[i]/sumX)
-    end
+    # credits: https://stats.stackexchange.com/questions/79454/softmax-layer-in-a-neural-network?newreg=d1e89b443dd346ae8bccaf038a944221
+    m,n =size(l.x)
 
-    for i = 1: ndims(X)
-        t = z[i]
-        w = zeros(ndims(X))
-        w[i] = 1
-        w = w .- z
-        u[:,i] = t * w
+    ly = Array{Float64}(n)
```

Comment:
`ly` should serves as a cache to hold array of length `n` during the construction of `jacobian` at each `batch`. Resolved by adding `l.ly` as `Softmax`'s instance and allocate it only once during `init`.